core/src/main/java/org/apache/accumulo/core/spi/compaction/SimpleCompactionDispatcher.java

```diff
@@ -72,19 +72,17 @@
   public void init(InitParameters params) {
     services = new EnumMap<>(CompactionKind.class);
 
-    var defaultService = CompactionDirectives.builder().build();
+    CompactionsDirectiveImpl defaultService = CompactionsDirectiveImpl.DEFAULT;
 
     if (params.getOptions().containsKey("service")) {
-      defaultService =
-          CompactionDirectives.builder().setService(params.getOptions().get("service")).build();
+      defaultService.setService(params.getOptions().get("service"));
     }
 
     for (CompactionKind ctype : CompactionKind.values()) {
       String service = params.getOptions().get("service." + ctype.name().toLowerCase());
-      if (service == null)
-        services.put(ctype, defaultService);
-      else
-        services.put(ctype, CompactionDirectives.builder().setService(service).build());
+      if (service != null)
+        defaultService.setService(service);
```

Comment:
This is changing a static object used by many different dispatcher instance.  This would cause a config change on one table to affect all other tables.  Also the builder used to create immutable objects.

Reply:
It was not clear to me that this class was immutable, since it contained multiple `setter` methods. I think a POJO with final members set in the constructor would be a lot more clear.

Reply:
Changing CompactionDirectives from an interface w/ a builder to a POJO with a single constructor and final instance vars would be an alternative.  I intentionally chose not to take that route because I have seen too many things in Accumulo end up with lots over overloaded constructors and methods over time.  The builder pattern is a good way to avoid this.  If nothing is ever added the POJO would be better overall, but I like I mentioned with the ScanDirectives I never could have imagined adding the caching directives in 2.1.0 beforehand.     Also I don't think there is much complexity currently from the perspective of someone using the current incarnation of CompactionDirectives.  There is certainly a lot of complexity in the impl class and some of that comes from trying to avoid object allocations.

Reply:
Another alternative to the POJO would be an interface w/ static methods that create an instance.  This has the same potential problem as constructors, there could be lots of overloaded methods in the future.  I think one nice thing about an interface w/ a static method is that if in the future multiple things are going to be added then a static builder() method could be added.  

Something like the following in 2.1.0

```java
public interface CompactionDirectives {

  /**
   * @return The service where a compaction should run.
   */
  CompactionServiceId getService();
 
  static CompactionDirectives from(CompactionServiceId){...}
}
```

Then in 2.2.0 we think of something else we want to return.

```java
public interface CompactionDirectives {

  /**
   * @return The service where a compaction should run.
   */
  CompactionServiceId getService();
 
  static CompactionDirectives from(CompactionServiceId){}
 /**
  * @since 2.2.0
  */
  TypeX getX();

 /**
  * @since 2.2.0
  */
 static CompactionDirectives from(CompactionServiceId csid, TypeX x){}
}
```

Then in 2.3.0 we think of 5 things we want to add and at that point just add a builder.

```java
public interface CompactionDirectives {

  /**
   * @return The service where a compaction should run.
   */
  CompactionServiceId getService();
 
  static CompactionDirectives from(CompactionServiceId){}
 /**
  * @since 2.2.0
  */
  TypeX getX();

 /**
  * @since 2.2.0
  */
 static CompactionDirectives from(CompactionServiceId csid, TypeX x){}

 /**
  * @since 2.3.0
  */
  interface Builder {...}

 /**
  * @since 2.3.0
  */
  static Builder builder(); 
}
```

Reply:
I am just trying to figure out if there is a way to make your highly evolved API code easier to understand for new developers. After playing around with it all afternoon, I still don't have a way to do that. I just worry because I _know_ that I have looked at the ScanDispatcher code before and yet, it still took me a few minutes to comprehend it again. Oh well, I digress.

Reply:
> I am just trying to figure out if there is a way to make your highly evolved API code easier to understand for new developers.

Easier to understand for an Accumulo user writing a compaction dispatcher or easier to understand for an Accumulo developer working on internal code?  For the latter, CompactionDirectivesImpl is complicated.. I may have gotten carried away trying to avoid object allocations which I think is at the root of the complexity.  For the former, if we want to clean up the user facing SPI it would good to do that before 2.1 is released.

Reply:
To simplify CompactionDirectivesImpl w/o changing the current SPI, could create two classes like the followin.

```java
class CompactionDirectiveBuilder implements Builder {

  private CompactionServiceId service;
  
  @Override
  public Builder setService(CompactionServiceId service) {
    this.service = service;
    return this;
  }

  @Override
  public Builder setService(String compactionServiceId) {
    this.service = CompactionServiceId.of(compactionServiceId);
    return this;
  }

  @Override
  public CompactionDirectives build() {
    return new CompactionsDirectiveImpl(service);
  }
  
}

class CompactionsDirectiveImpl implements CompactionDirectives {

  private final CompactionServiceId service;
  
  public CompactionsDirectiveImpl(CompactionServiceId service) {
    this.service = service;
  }

  @Override
  public CompactionServiceId getService() {
    return service;
  }

  @Override
  public String toString() {
    return "service=" + service;
  }
}
```

Reply:
> Easier to understand for an Accumulo user writing a compaction dispatcher or easier to understand for an Accumulo developer working on internal code?

I think both would be good goals. But I was first targeting the latter. 

I think splitting the two implementations is a big help. I will replace the changes on this PR with the two separate classes.